python/pyspark/sql/types.py

```diff
@@ -619,7 +621,8 @@ def fromInternal(self, obj):
             # it's already converted by pickler
             return obj
         if self._needSerializeAnyField:
-            values = [f.fromInternal(v) for f, v in zip(self.fields, obj)]
+            values = [f.fromInternal(v) if n else v
```

Comment:
Ah, I see. This can be recursive and per-record and we avoid here by pre-computing. I see. That makes much sense.

Reply:
Let's describe this in more details and add some numbers (and in your other PRs too).